### PR TITLE
feat(cli): make osm version print remote version

### LINKED ISSUE #1720: Add check in the OSM CLI for available updates / upgrades / new versions
This issue is to implement an OSM CLI command, which performs an upgrade of the control plane (and data plane if necessary) to the latest available version. For example: `osm check-update` could output `Current installed version: X;  Available new version: Y`

The design of this feature must ensure that Kubernetes clusters with older versions of the SMI CRDs continue to operate untinterrupted.

For instance: a cluster with version 1 of SMI and version 1 of OSM is upgraded to version 2 of OSM.  Version 2 of OSM must continue to work with version 1 of SMI.

Unanswered questions:
 - how many legacy SMI CRD versions is the latest OSM control plane going to support?

**Scope (please mark with X where applicable)**
- New Functionality          [ ]
- Install                    [ ]
- SMI Traffic Access Policy  [ ]
- SMI Traffic Specs Policy   [ ]
- SMI Traffic Split Policy   [ ]
- Permissive Traffic Policy  [ ]
- Ingress                    [ ]
- Egress                     [ ]
- Envoy Control Plane        [ ]
- CLI Tool                   [ ]
- Metrics                    [ ]
- Certificate Management     [ ]
- Sidecar Injection          [ ]
- Logging                    [ ]
- Debugging                  [ ]
- Tests                      [ ]
- CI System                  [ ]
- Project Release            [ ]

**Possible use cases**

### PR DESCRIPTION
**Pull request by johnsonshi**
**Description**:

Modify `osm version` cli command to print:
(1) osm version installed remotely in mesh name
and namespace,
(2) local osm CLI version,
(3) latest available release version.

Add `remote` boolean flag to `osm version`, which
can be used to suppress checking the remote
osm version installed in the mesh.

Resolves #1720.

Signed-off-by: Johnson Shi <Johnson.Shi@microsoft.com>

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [X] |
| Documentation              | [ ] |
| Install                    | [ ] |
| CLI Tool                   | [X] |
| Certificate Management     | [ ] |
| Control Plane              | [ ] |
| Ingress                    | [ ] |
| Egress                     | [ ] |
| Networking                 | [ ] |
| Observability              | [ ] |
| SMI Policy                 | [ ] |
| Sidecar Injection          | [ ] |
| Security                   | [ ] |
| Upgrade                    | [X] |
| Tests                      | [ ] |
| CI System                  | [ ] |
| Demo                       | [ ] |
| Performance                | [ ] |
| Other                      | [ ] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project? No.
    -   Did you notify the maintainers and provide attribution? N/A

1. Is this a breaking change? No.

### CODE REVIEW

File: cmd/cli/mesh_check_upgrade.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+	"github.com/openservicemesh/osm/pkg/constants"
+)
+
+const (
+	osmReleaseRepoOwner = "openservicemesh"
+	osmReleaseRepoName = "osm"
+)
+
+const meshCheckUpgradeDesc = `
```
Comment (shashankram) [inline cmd/cli/mesh_check_upgrade.go:21]: Can this be integrated into the `osm version` command instead where it displays the existing version as well as new ones available?

**Review by shashankram — COMMENTED** [review]
Hey @johnsonshi, this is a useful change. Would you mind providing a brief description in the commit message as per the commit message guidelines, thanks.

image

File: cmd/cli/mesh_check_upgrade.go (hunk 1 above)
Comment (johnsonshi) [reply cmd/cli/mesh_check_upgrade.go:21]: Hmmm that's a better experience indeed! I'll merge the two commands together since it makes sense!

**Comment by johnsonshi** [comment]
@shashankram // @draychev as advised by Shashank, I decided to integrate both commands under `osm version`.
`osm version` now prints:
(1) the OSM version installed in the mesh name and namespace,
(2) the local OSM CLI version,
(3) the latest OSM release version available.

Additionally, `osm version` now has a new boolean flag `remote`. The `remote` flag is used by the user to suppress checking/printing remote osm version information. This makes the `osm version` experience in line with `istioctl version` experience, which prints the cli version + remote mesh version, but with a `remote` flag option to suppress checking remote version. See below.

`istioctl version`:
```
~ » istioctl version --help                                                                                                                                                                                        johnsonshi@OSOC-DLAPT-21
Prints out build version information

Usage:
  istioctl version [flags]

Flags:
  -h, --help              help for version
  -o, --output string     One of 'yaml' or 'json'.
      --remote            Use --remote=false to suppress control plane check
  -r, --revision string   Control plane revision
  -s, --short             Use --short=false to generate full version information

Global Flags:
      --context string          The name of the kubeconfig context to use
  -i, --istioNamespace string   Istio system namespace (default "istio-system")
  -c, --kubeconfig string       Kubernetes configuration file
  -n, --namespace string        Config namespace
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~ » istioctl version                                                                                                                                                                                               johnsonshi@OSOC-DLAPT-21
no running Istio pods in "istio-system"
1.10.0
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~ » istioctl version --remote                                                                                                                                                                                      johnsonshi@OSOC-DLAPT-21
no running Istio pods in "istio-system"
1.10.0
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~ » istioctl version --remote=false                                                                                                                                                                                johnsonshi@OSOC-DLAPT-21
1.10.0
```


`osm version`:
```
~/dev/osm(resolve-1720*) » ./bin/osm version                                                                                                                                                                       johnsonshi@OSOC-DLAPT-21
No osm mesh installation found with mesh name [osm] and namespace [osm-system].
Local CLI Version: dev; Commit: eae0ecafcc7ea3384997a31c5eeccc7cb57b2997; Date: 2021-06-13-14:06
Latest Available Version: v0.8.4.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/dev/osm(resolve-1720*) » ./bin/osm version --remote=false                                                                                                                                                        johnsonshi@OSOC-DLAPT-21
Local CLI Version: dev; Commit: eae0ecafcc7ea3384997a31c5eeccc7cb57b2997; Date: 2021-06-13-14:06
Latest Available Version: v0.8.4.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/dev/osm(resolve-1720*) » ./bin/osm version --remote=true                                                                                                                                                         johnsonshi@OSOC-DLAPT-21
No osm mesh installation found with mesh name [osm] and namespace [osm-system].
Local CLI Version: dev; Commit: eae0ecafcc7ea3384997a31c5eeccc7cb57b2997; Date: 2021-06-13-14:06
Latest Available Version: v0.8.4.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/dev/osm(resolve-1720*) » osm install                                                                                                                                                                             johnsonshi@OSOC-DLAPT-21
OSM installed successfully in namespace [osm-system] with mesh name [osm]
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/dev/osm(resolve-1720*) » osm install --mesh-name osm-2 --osm-namespace ns-2                                                                                                                                      johnsonshi@OSOC-DLAPT-21
OSM installed successfully in namespace [ns-2] with mesh name [osm-2]
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/dev/osm(resolve-1720*) » ./bin/osm version                                                                                                                                                                       johnsonshi@OSOC-DLAPT-21
Mesh Name: osm. Mesh Namespace: osm-system. Mesh Version: v0.8.4.
Local CLI Version: dev; Commit: eae0ecafcc7ea3384997a31c5eeccc7cb57b2997; Date: 2021-06-13-14:06
Latest Available Version: v0.8.4.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/dev/osm(resolve-1720*) » ./bin/osm version --mesh-name osm-2 --osm-namespace ns-2                                                                                                                                johnsonshi@OSOC-DLAPT-21
Mesh Name: osm-2. Mesh Namespace: ns-2. Mesh Version: v0.8.4.
Local CLI Version: dev; Commit: eae0ecafcc7ea3384997a31c5eeccc7cb57b2997; Date: 2021-06-13-14:06
Latest Available Version: v0.8.4.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/dev/osm(resolve-1720*) » ./bin/osm version --mesh-name XXX --osm-namespace XXX                                                                                                                                   johnsonshi@OSOC-DLAPT-21
No osm mesh installation found with mesh name [XXX] and namespace [XXX].
Local CLI Version: dev; Commit: eae0ecafcc7ea3384997a31c5eeccc7cb57b2997; Date: 2021-06-13-14:06
Latest Available Version: v0.8.4.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/dev/osm(resolve-1720*) » ./bin/osm version --help                                                                                                                                                                johnsonshi@OSOC-DLAPT-21

This command prints out the following version information:
(1) the OSM version installed in the mesh name and namespace,
(2) the local OSM CLI version,
(3) the latest OSM release version available.

The mesh to check for version information is identified
by its mesh name and namespace.
Use the --mesh-name and --osm-namespace flags to specify the mesh to
check for version information.

Usage:
  osm version [flags]

Examples:

# Print the osm version in the mesh "osm1" and namespace "osm-ns1".
# Also print the local CLI version and the latest release version available.
osm version --mesh-name osm1 --osm-namespace osm-ns1


Flags:
  -h, --help               help for version
      --mesh-name string   Name of the mesh to check for version information (default "osm")
      --remote             Use remote=false to skip checking osm version installed in the remote mesh. (default true)

Global Flags:
      --osm-namespace string   namespace for osm control plane (default "osm-system")
```

File: cmd/cli/version.go [hunk 2]
```diff
@@ -1,32 +1,138 @@
...
+
+		if err != nil {
+			_, _ = fmt.Fprint(c.out, err.Error()+"\n")
+		} else {
+			// get mesh version currently installed
+			meshVersion := osmControllerDeployment.ObjectMeta.Labels[constants.OSMAppVersionLabelKey]
+			if meshVersion == "" {
+				meshVersion = "Unknown"
+			}
+			_, _ = fmt.Fprintf(c.out, "Mesh Name: %s. Mesh Namespace: %s. Mesh Version: %s.\n", c.meshName, meshNamespace, meshVersion)
```
Comment (jaellio) [inline cmd/cli/version.go:96]: nit: The local CLI version information is separated by semi-colons. To be consistent the remote version information could also be separated by semi-colons.

**Review by jaellio — APPROVED** [review]

File: cmd/cli/version.go [hunk 3]
```diff
@@ -1,32 +1,138 @@
...
+
+		if err != nil {
+			_, _ = fmt.Fprint(c.out, err.Error()+"\n")
+		} else {
+			// get mesh version currently installed
+			meshVersion := osmControllerDeployment.ObjectMeta.Labels[constants.OSMAppVersionLabelKey]
+			if meshVersion == "" {
+				meshVersion = "Unknown"
+			}
+			_, _ = fmt.Fprintf(c.out, "Mesh Name: %s; Mesh Namespace: %s; Mesh Version: %s\n", c.meshName, meshNamespace, meshVersion)
```
Comment (allenlsy) [inline cmd/cli/version.go:96]: Why not drop all return values? Is this not allowed by the linter?

**Comment by allenlsy** [comment]
Should we remove `cmd/cli/chart.tgz` file from this commit?

**Review by shashankram — CHANGES_REQUESTED** [review]
The binary file should be removed.

File: cmd/cli/util_test.go [hunk 4]
```diff
@@ -125,6 +126,28 @@ var _ = Describe("Test getting pretty printed output of a list of meshes", func(
 	})
 })
 
+var _ = Describe("Test getting mesh controller metadata", func() {
```
[the remark this answers is not in the document]
Comment (shashankram) [reply cmd/cli/util_test.go:129]: Yes, let's write new tests using Go's `testing` pkg.

**Comment by johnsonshi** [comment]
This PR needs to be reworked due to design changes and needing additional work. Closing this PR.

**Pull request closed without merging**